### DISCUSSION TITLE
Running async events without button triggers.

### DISCUSSION THREAD

**Discussion by rohitsathish** [zauberzeug/nicegui#2691] [Q&A] [answered]
### Question

I have an async function ```load_data1``` that I would like to execute in the script. But I would like this to be run each time the site is loaded, so I guess ```app.on_connect``` is the right function. But it doesn't seem to do the job. Also, there are a lot of examples to run functions with  ```on_click``` in the documentation, but I'd like to know if there are other backend methods to trigger events.  A minimal example below:

```python

from nicegui import ui, run, app

async def load_data1():
    return "hello"

def df_layout():

    async def show_df():
        app.storage.user['df'] = await load_data1()
        ui.notify(app.storage.user['df'])
        ui.label(app.storage.user['df'])

    app.on_connect(show_df)
    #ui.button('Show', on_click=show_df)


@ui.page("/")
async def main_page():
    df_layout()

ui.run(storage_secret='test', port=8081)
```

Other issues have mentioned how the error reports could be improved as it doesn't reference the actual problem :
```RuntimeError: app.storage.user needs a storage_secret passed in ui.run()```

**Comment by rodja** [comment, marked as answer]
`app.on_connect` is only meant for global hooks. If you want to execute code after after connection is established on a individual page you should use `await client.connected()`:

```py
from nicegui import ui, context, app

async def load_data1():
    return "hello"

async def df_layout():
    async def show_df():
        app.storage.user['df'] = await load_data1()
        ui.notify(app.storage.user['df'])
        ui.label(app.storage.user['df'])

    await context.get_client().connected()
    await show_df()
    # ui.button('Show', on_click=show_df)


@ui.page("/")
async def main_page():
    await df_layout()

ui.run(storage_secret='test')
```

**Comment by rodja** [comment]
This can also be simplified:

```py
@ui.page("/")
async def main_page(client: Client):
    await client.connected()
    app.storage.user['df'] = await load_data1()
    ui.notify(app.storage.user['df'])
    ui.label(app.storage.user['df'])
```
